### DISCUSSION TITLE
Can moose implement the FE2 algorithm?

### DISCUSSION THREAD

**Discussion by stonebeta** [idaholab/moose#26991] [Q&A Modules: Solid mechanics]
like this
![image](https://github.com/idaholab/moose/assets/85184413/1ec9ad23-3c2f-4d7b-be37-9c48de66d637)

**Comment by GiudGiud** [comment]
Hello

This seems fine to do with a MultiApp. see the tutorial there
https://mooseframework.inl.gov/getting_started/examples_and_tutorials/tutorial02_multiapps/presentation/index.html#/

There s probably examples in moose that are similar?
What do you think of this one?
https://mooseframework.inl.gov/modules/phase_field/Mechanics_Coupling.html

Guillaume

**Reply by stonebeta** [reply]
In the process of Newton iteration, how to synchronize the average stress of the child-app to the quadrature points of the parent-app, and how to apply the strain of each quadrature point of the parent-app to the constraint functions of the homogenization system used in the child-apps?

**Reply by GiudGiud** [reply]
Child app and parent app synchronize through fixed point iterations. 
You'll use transfers to pass fields in the parent app to the child app. Is the constraint function constant over each child app?

**Reply by stonebeta** [reply]
I believe that the strain of the parent app should be passed to the parameter `target` of `HomogenizationConstraint` to control the child app (RVE problem). However, the parameter `target`  accepts a function. Is it possible to pass the variables of the parent app to the function of the child app?

**Reply by GiudGiud** [reply]
No functions cannot be filled with variable data. If the function can be spatially constant, I would transfer from the main app to postprocessors in the subapp. Then a ParsedFunction can be made equal to a postprocessor

**Reply by hugary1995** [reply]
@xiekai-mc There is a crude implementation of FE2 here: https://github.com/peiyichen1995/fe2 which hopefully can give you some idea on how it works.
See [`FE2PK1Stress`](https://github.com/peiyichen1995/fe2/blob/main/src/materials/FE2PK1Stress.C) and the use of `QuadraturePointMultiApp`

I believe @abarun22 is doing something along the same lines. He may be able to offer some suggestions.

**Reply by stonebeta** [reply]
Thank you for your reply. I've looked into the code of [peiyichen1995/fe2](https://github.com/peiyichen1995/fe2).
In the project ,  `_fe_problem.execMultiApps` is utilized within `computeQpPK1Stress()` to run a subapp. It use a UserObject to temporarily store the inputs and outputs of the subapp, and compute stress[_qp] and jacobian[_qp] through perturbation method. The subapp is steady and does not retain historical data.
My idea is, if using QuadraturePointMultiApp, it should be within `computeProperties()` rather than `computeQpPK1Stress()` to execute subapps using `_fe_problem.execMultiApps`.  If perturbation is also utilized to compute the jacobian, the subapps would need to be reset, reverting them to their state prior to the time step calculation. This may require using `_fe_problem.restoreMultiApps`.
However, the first issue I encountered now is that when using QuadraturePointMultiApp and running `_fe_problem.execMultiApps` within `computeProperties()`, an error occurs.
```
Assertion `!Threads::in_threads' failed
PerfGraph timing cannot be used within threaded sections
at /home/slime/projects/moose/framework/src/problems/FEProblemBase.C, line 4908
```

**Reply by GiudGiud** [reply]
hello

Are you using threads?
executing multiapps inside computeProperties is going to be expensive. If you still want to try, you will have to not use threads and either:
- run this code in opt mode so the assert does not get triggered
- OR comment out this line of FEProblem

Guillaume
